protocol/contracts/beanstalk/sun/SeasonFacet/Distribution.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {C} from "contracts/C.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {AppStorage, Storage} from "contracts/beanstalk/AppStorage.sol";
+import {ShipmentPlan, Receiving} from "./Receiving.sol";
+
+/**
+ * @title Distribution
+ * @author funderbrker
+ * @notice Handles shipping of new Bean mints.
+ */
+contract Distribution is Receiving {
+    using SafeCast for uint256;
+
+    /**
+     * @notice Emitted during Sunrise when Beans are distributed to the Field, the Silo, and Fertilizer.
+     * @param season The Season in which Beans were distributed.
+     * @param shipmentRoutes The routes defining where to distribute.
+     * @param shipmentAmounts The amount of Beans distributed to each route.
+     */
+    event Ship(
+        uint32 indexed season,
+        Storage.ShipmentRoute[] shipmentRoutes,
+        uint256[] shipmentAmounts
+    );
+
+    //////////////////// REWARD BEANS ////////////////////
+
+    /**
+     * @notice Mints and distributes Beans across all active shipping routes.
+     * @param beansToShip The total number of Beans to distribute.
+     */
+    function ship(uint256 beansToShip) internal {
+        C.bean().mint(address(this), beansToShip);
+
+        Storage.ShipmentRoute[] memory shipmentRoutes = s.shipmentRoutes;
+        ShipmentPlan[] memory shipmentPlans = new ShipmentPlan[](shipmentRoutes.length);
+        uint256[] memory shipmentAmounts = new uint256[](shipmentRoutes.length);
+        uint256 totalPoints;
+        (shipmentPlans, totalPoints) = getShipmentPlans(shipmentRoutes);
+
+        // May need to calculate individual stream rewards multiple times, since
+        // they are dependent on each other and each has a cap.
+        for (uint256 i; i < shipmentRoutes.length; i++) {
```

Comment:
Is this implmented without the loop in the case that a cap is hit currently?

Reply:
I am unsure what you mean.
If no cap is hit in the first pass it will break out without performing unnecessary loops

Reply:
Got it, this is great! Realized that the worst case would be `O(n^2)`, which checks out here. Would the shipment plans need to be re-calculated in the loop, given they are dependent on each other? Currently the current system redistributes the extra beans proportional to the new points %. Ideally the system would have more flexibility in controlling its distribution. 

Imagine a hypothetical setup where we have a silo, active field, barn, and legacy field, with the distribution of 45%, 45%, 5%, 5%. Respectively. Say the beanstalk wants to setup distribution such that: 
    - if the barn is paid off, extra bean seignorage goes to the legacy field. 
    - if the legacy field is paid off, extra bean seignorage goes to the barn. 
    - if the active field is paid off, extra bean seignorage goes to the silo.

With the current system, It doesn't seem possible to implement static numbers to fit this functionality. For example if the active field was paid off, the new distribution would be 80%, 10%, 10% respectively.

Reply:
That functionality is not supported by this design. Once a system is paid off it will cap itself or set its points to 0, which will result in all other output streams receiving a larger portion of the total bean mints.

It is not clear how a fully circularly dependent system could be designed, but certain features could likely be achieved. However, what your describing is out of scope atm.